pipelines/salesforce/__init__.py

```diff
@@ -0,0 +1,226 @@
+            state["last_replication_key"] = max(
+                state.get("last_replication_key", ""),
+                page[-1][replication_key],
+            )
+
+    print(f"Done with {sobject} extraction. {n_records} records extracted.")
+    if state and n_records:
+        print("New state:", state)
+
+
+def sobject_factory(
+    name: str,
+    sobject: str,
+    client: Salesforce,
+    write_disposition: str = "replace",
+    replication_key: t.Optional[str] = None,
+    deferrable: bool = False,
+    fields: t.Optional[t.List[str]] = None,
+) -> t.Callable[[], t.Iterable[dict]]:
+    """Return a resource for a Salesforce sobject.
+
+    Args:
+        name: The name of the resource. This affects the table name in the warehouse.
+        sobject: The Salesforce object API name.
+        client: A Salesforce client.
+        write_disposition: The write disposition for the resource.
+        replication_key: The field to use for incremental replication.
+        deferrable: Whether the resource should be deferrable. We recommend not
+            deferring large resources as it could cause memory issues.
+        fields: A list of fields to extract. If not specified, all fields
+            will be extracted.
+    """
+
+    @dlt.resource(name=name, write_disposition=write_disposition)
+    def sobject_resource():
+        nonlocal write_disposition, replication_key
+
+        if replication_key and write_disposition == "append":
+            state: dict = dlt.state().setdefault(name, {})
+        else:
+            replication_key = state = None
+
+        if deferrable:
+            yield dlt.defer(
+                lambda: [
+                    record
+                    for page in get_records(
+                        client, sobject, state, replication_key, fields
+                    )
+                    for record in page
+                ]
+            )
+        else:
+            yield from get_records(client, sobject, state, replication_key, fields)
+
+    return sobject_resource
+
+
+@dlt.source(name="salesforce")
+def salesforce_source(
+    username: str = dlt.secrets.value,
+    password: str = dlt.secrets.value,
+    security_token: str = dlt.secrets.value,
+):
+    client = Salesforce(username, password, security_token)
+    # You can add more sobjects here, custom ones too
+    # See standard objects here:
+    # https://developer.salesforce.com/docs/atlas.en-us.object_reference.meta/object_reference/sforce_api_objects_list.htm
+    # See the `sobject_factory` function docstring for info on the arguments
+    return (
+        sobject_factory(name="sf_user", sobject="User", client=client, deferrable=True),
+        sobject_factory(
+            name="user_role", sobject="UserRole", client=client, deferrable=True
+        ),
+        sobject_factory(
+            name="opportunity",
+            sobject="Opportunity",
+            client=client,
+            write_disposition="append",
```

Comment:
looks like all the append resources are on modification date. this means that those objects can be retrieved many times and we'll generate duplicates. IMO we need to change it to `merge`
https://help.salesforce.com/s/articleView?id=000387261&type=1

Reply:
done, do I need to specify the primary key if I use merge_ it seems to accept it without